linkedin/eyeglass
# feat: Add config for conditional Embroider builds.

BREAKING BUGFIX

Requires apps to declare if they are using Embroider to build
via config option in `ember-cli-build.js`.

## More Details

Apps may opt to build using Embroider conditionally. This means they'll have `@embroider/core` included in their dependencies but may _not_ be building using Embroider. This can cause extraneous files to end up in the output tree, which can cascade into failures in downstream addons (such as `ember-cli-terser`).

This adds a config that apps **must** use to declare if they're using Embroider conditionally. This solely controls whether `ember-cli-eyeglass` builds using "Embroider mode" or not. (The previous code of detecting what dependencies the app has added is removed.)

## How to Reproduce the Bug

1. In `eyeglass-test-app`, add `ember-cli-terser` and `@embroider/core` to the build. Do _not_ make any changes to `ember-cli-build.js`.
2. Add a Glimmer component to the addon. Be sure to use at least one decorator, such as `@tracked` or `@action`.
3. Run `ember build --prod`.
4. `ember-cli-terser` will fail the build when it trips across the uncompiled decorator in the code.

If you run with `BROCCOLI_DEBUG="*"`, you'll note a bunch of files in the `output` dir for `ember-cli-eyeglass` that shouldn't be there. These eventually end up in the `input` to `ember-cli-terser`.

## How to Verify Fix

1. In `eyeglass-test-app`, after applying the above steps, add `embroiderEnabled: false` to the eyeglass config in `ember-cli-build.js`.
2. Run `ember build --prod`.
3. It should succeed! 🎉 

When running with Broccoil Debug, the `output` dir should only include compiled CSS files and other CSS files that didn't need to be compiled, nested in an `assets` dir.

I also verified this fix by running `yarn test` on `ember-cli-eyeglass`.
